AssertionError when using label_cols in to_tf_dataset

## Describe the bug
An incorrect `AssertionError` is raised when using `label_cols` in `to_tf_dataset` and the label's key name is `label`.

The assertion is in this line:
https://github.com/huggingface/datasets/blob/2.4.0/src/datasets/arrow_dataset.py#L475

## Steps to reproduce the bug
```python
from datasets import load_dataset
from transformers import DefaultDataCollator

dataset = load_dataset('glue', 'mrpc', split='train')

tf_dataset = dataset.to_tf_dataset(
    columns=["sentence1", "sentence2", "idx"],
    label_cols=["label"],
    batch_size=16,
    collate_fn=DefaultDataCollator(return_tensors="tf"),
)
```

## Expected results
No assertion error.

## Actual results
```
AssertionError: in user code:

    File "/opt/conda/lib/python3.8/site-packages/datasets/arrow_dataset.py", line 475, in split_features_and_labels  *
        assert set(features.keys()).union(labels.keys()) == set(input_batch.keys())
```

## Environment info
<!-- You can run the command `datasets-cli env` and copy-and-paste its output below. -->
- `datasets` version: 2.4.0
- Platform: Linux-4.18.0-305.45.1.el8_4.ppc64le-ppc64le-with-glibc2.17
- Python version: 3.8.13
- PyArrow version: 7.0.0
- Pandas version: 1.4.3

--- a/src/datasets/arrow_dataset.py
+++ b/src/datasets/arrow_dataset.py
@@ -260,7 +260,7 @@ class TensorflowDatasetMixin:
         if batch_size is not None:
             batch_size = min(len(dataset), batch_size)
         test_batch_size = min(len(dataset), 2)
-        
+
         if cols_to_retain is None or not set(cols_to_retain) & {"label", "label_ids", "labels"}:
             auto_rename_labels = False
         if auto_rename_labels:

--- a/src/datasets/arrow_dataset.py
+++ b/src/datasets/arrow_dataset.py
@@ -225,7 +225,6 @@ class TensorflowDatasetMixin:
         cols_to_retain: Optional[List[str]] = None,
         batch_size: Optional[int] = None,
         num_test_batches: int = 200,
-        auto_rename_labels: bool = True,
     ):
         """Private method used by `to_tf_dataset()` to find the shapes and dtypes of samples from this dataset
            after being passed through the collate_fn. Tensorflow needs an exact signature for tf.numpy_function, so
@@ -243,8 +242,6 @@ class TensorflowDatasetMixin:
             batch_size (:obj:`int`, optional): The size of batches loaded from the dataset. Used for shape inference.
                 Can be None, which indicates that batch sizes can be variable.
             num_test_batches (:obj:`int`): The number of batches to load from the dataset for shape inference.
-            auto_rename_labels (:obj:`bool`): Whether to automatically rename labels from "label" or "label_ids"
-                to match `transformers`. In future, we will probably move this to the `transformers` library itself.
 
         Returns:
             :obj:`dict`: Dict mapping column names to tf.Tensorspec objects
@@ -261,9 +258,7 @@ class TensorflowDatasetMixin:
             batch_size = min(len(dataset), batch_size)
         test_batch_size = min(len(dataset), 2)
 
-        if cols_to_retain is None or not set(cols_to_retain) & {"label", "label_ids", "labels"}:
-            auto_rename_labels = False
-        if auto_rename_labels:
+        if cols_to_retain is not None:
             cols_to_retain = list(set(cols_to_retain + ["label_ids", "label", "labels"]))
 
         test_batches = []
@@ -330,7 +325,6 @@ class TensorflowDatasetMixin:
         collate_fn_args: Optional[Dict[str, Any]] = None,
         label_cols: Optional[Union[str, List[str]]] = None,
         prefetch: bool = True,
-        auto_rename_labels: bool = True,
     ):
         """Create a tf.data.Dataset from the underlying Dataset. This tf.data.Dataset will load and collate batches from
         the Dataset, and is suitable for passing to methods like model.fit() or model.predict(). The dataset will yield
@@ -355,12 +349,6 @@ class TensorflowDatasetMixin:
             prefetch (:obj:`bool`, default ``True``): Whether to run the dataloader in a separate thread and maintain
                 a small buffer of batches for training. Improves performance by allowing data to be loaded in the
                 background while the model is training.
-            auto_rename_labels (:obj:`bool`, default ``True``): Whether to automatically handle the case where labels
-                are renamed by data collators, particularly those used in the `transformers` library. Label columns
-                are usually named "label" in `datasets`, but `transformers` models use "labels" instead. This renaming
-                is usually handled by the data collator. This argument is usually safe to leave as `True`, even when
-                not using `transformers` models or data collators, but we provide the option to disable it as it may
-                create strange edge cases when you want to preserve multiple columns that have label-like names.
 
         Returns:
             :class:`tf.data.Dataset`
@@ -411,18 +399,12 @@ class TensorflowDatasetMixin:
         else:
             cols_to_retain = None  # Indicates keeping all valid columns
             columns = []
-        if cols_to_retain is None or not (set(cols_to_retain) & {"label", "labels", "label_ids"}):
-            auto_rename_labels = False
 
         if self.format["type"] != "custom":
             dataset = self.with_format("numpy")
         else:
             dataset = self
 
-        # Following the logic in `transformers.Trainer`, we do not drop `label_ids` or `label` even if they
-        # are not in the list of requested columns when auto_rename_labels is set, because the collator may rename them
-        # This might work better if moved to a method attached to our transformers Model objects, but doing so
-        # could break backward compatibility
         # TODO(Matt, QL): deprecate the retention of label_ids and label
 
         output_signature, columns_to_np_types = dataset._get_output_signature(
@@ -431,10 +413,9 @@ class TensorflowDatasetMixin:
             collate_fn_args=collate_fn_args,
             cols_to_retain=cols_to_retain,
             batch_size=batch_size if drop_remainder else None,
-            auto_rename_labels=auto_rename_labels,
         )
 
-        if auto_rename_labels and "labels" in output_signature:
+        if "labels" in output_signature:
             if ("label_ids" in columns or "label" in columns) and "labels" not in columns:
                 columns = [col for col in columns if col not in ["label_ids", "label"]] + ["labels"]
             if ("label_ids" in label_cols or "label" in label_cols) and "labels" not in label_cols:
@@ -455,7 +436,7 @@ class TensorflowDatasetMixin:
             else:
                 batch = dataset[indices]
 
-            if cols_to_retain is not None and auto_rename_labels:
+            if cols_to_retain is not None:
                 batch = {
                     key: value
                     for key, value in batch.items()

--- a/src/datasets/arrow_dataset.py
+++ b/src/datasets/arrow_dataset.py
@@ -225,6 +225,7 @@ class TensorflowDatasetMixin:
         cols_to_retain: Optional[List[str]] = None,
         batch_size: Optional[int] = None,
         num_test_batches: int = 10,
+        auto_rename_labels: bool = True,
     ):
         """Private method used by `to_tf_dataset()` to find the shapes and dtypes of samples from this dataset
            after being passed through the collate_fn. Tensorflow needs an exact signature for tf.numpy_function, so
@@ -241,6 +242,9 @@ class TensorflowDatasetMixin:
                 `collate_fn`.
             batch_size (:obj:`int`, optional): The size of batches loaded from the dataset. Used for shape inference.
                 Can be None, which indicates that batch sizes can be variable.
+            num_test_batches (:obj:`int`): The number of batches to load from the dataset for shape inference.
+            auto_rename_labels (:obj:`bool`): Whether to automatically rename labels from "label" or "label_ids"
+                to match `transformers`. In future, we will probably move this to the `transformers` library itself.
 
         Returns:
             :obj:`dict`: Dict mapping column names to tf.Tensorspec objects
@@ -258,17 +262,17 @@ class TensorflowDatasetMixin:
         else:
             batch_size = min(len(dataset), batch_size)
             test_batch_size = batch_size
+        if cols_to_retain is None or not set(cols_to_retain) & {"label", "label_ids", "labels"}:
+            auto_rename_labels = False
+        if auto_rename_labels:
+            cols_to_retain = list(set(cols_to_retain + ["label_ids", "label", "labels"]))
 
         test_batches = []
         for _ in range(num_test_batches):
             indices = sample(range(len(dataset)), test_batch_size)
             test_batch = dataset[indices]
             if cols_to_retain is not None:
-                test_batch = {
-                    key: value
-                    for key, value in test_batch.items()
-                    if key in cols_to_retain or key in ("label_ids", "label")
-                }
+                test_batch = {key: value for key, value in test_batch.items() if key in cols_to_retain}
             test_batch = [{key: value[i] for key, value in test_batch.items()} for i in range(test_batch_size)]
             test_batch = collate_fn(test_batch, **collate_fn_args)
             test_batches.append(test_batch)
@@ -327,6 +331,7 @@ class TensorflowDatasetMixin:
         collate_fn_args: Optional[Dict[str, Any]] = None,
         label_cols: Optional[Union[str, List[str]]] = None,
         prefetch: bool = True,
+        auto_rename_labels: bool = True,
     ):
         """Create a tf.data.Dataset from the underlying Dataset. This tf.data.Dataset will load and collate batches from
         the Dataset, and is suitable for passing to methods like model.fit() or model.predict(). The dataset will yield
@@ -351,6 +356,12 @@ class TensorflowDatasetMixin:
             prefetch (:obj:`bool`, default ``True``): Whether to run the dataloader in a separate thread and maintain
                 a small buffer of batches for training. Improves performance by allowing data to be loaded in the
                 background while the model is training.
+            auto_rename_labels (:obj:`bool`, default ``True``): Whether to automatically handle the case where labels
+                are renamed by data collators, particularly those used in the `transformers` library. Label columns
+                are usually named "label" in `datasets`, but `transformers` models use "labels" instead. This renaming
+                is usually handled by the data collator. This argument is usually safe to leave as `True`, even when
+                not using `transformers` models or data collators, but we provide the option to disable it as it may
+                create strange edge cases when you want to preserve multiple columns that have label-like names.
 
         Returns:
             :class:`tf.data.Dataset`
@@ -399,40 +410,61 @@ class TensorflowDatasetMixin:
                 raise ValueError("List of columns contains duplicates.")
             cols_to_retain = list(set(columns + label_cols))
         else:
-            cols_to_retain = None  # Indicates keeping all non-numerical columns
+            cols_to_retain = None  # Indicates keeping all valid columns
             columns = []
+        if cols_to_retain is None or not (set(cols_to_retain) & {"label", "labels", "label_ids"}):
+            auto_rename_labels = False
 
         if self.format["type"] != "custom":
             dataset = self.with_format("numpy")
         else:
             dataset = self
-        # If the user hasn't specified columns, give them all columns. This may break some data collators if columns
-        # are non-numeric!
 
-        # If drop_remainder is True then all batches will have the same size, so this can be included in the
-        # output shape. If drop_remainder is False then batch size can be variable, so that dimension should
-        # be listed as None
+        # Following the logic in `transformers.Trainer`, we do not drop `label_ids` or `label` even if they
+        # are not in the list of requested columns when auto_rename_labels is set, because the collator may rename them
+        # This might work better if moved to a method attached to our transformers Model objects, but doing so
+        # could break backward compatibility
+        # TODO(Matt, QL): deprecate the retention of label_ids and label
+
         output_signature, columns_to_np_types = dataset._get_output_signature(
             dataset,
             collate_fn=collate_fn,
             collate_fn_args=collate_fn_args,
             cols_to_retain=cols_to_retain,
             batch_size=batch_size if drop_remainder else None,
+            auto_rename_labels=auto_rename_labels,
         )
 
+        if auto_rename_labels and "labels" in output_signature:
+            if ("label_ids" in columns or "label" in columns) and "labels" not in columns:
+                columns = [col for col in columns if col not in ["label_ids", "label"]] + ["labels"]
+            if ("label_ids" in label_cols or "label" in label_cols) and "labels" not in label_cols:
+                label_cols = [col for col in label_cols if col not in ["label_ids", "label"]] + ["labels"]
+
+        for col in columns:
+            if col not in output_signature:
+                raise ValueError(f"Column {col} not found in dataset!")
+
+        for col in label_cols:
+            if col not in output_signature:
+                raise ValueError(f"Label column {col} not found in dataset!")
+
         def np_get_batch(indices):
-            # Following the logic in `transformers.Trainer`, we do not drop `label_ids` or `label` even if they
-            # are not in the list of requested columns, because the collator may rename them
-            # This might work better if moved to a method attached to our transformers Model objects, but doing so
-            # could break backward compatibility
-            # TODO(Matt, QL): deprecate the retention of label_ids and label
-            batch = dataset[indices]
-            if cols_to_retain is not None:
+            # Optimization - if we're loading a sequential batch, do it with slicing instead of a list of indices
+            if np.all(np.diff(indices) == 1):
+                batch = dataset[indices[0] : indices[-1] + 1]
+            else:
+                batch = dataset[indices]
+
+            if cols_to_retain is not None and auto_rename_labels:
                 batch = {
                     key: value
                     for key, value in batch.items()
-                    if key in cols_to_retain or key in ("label_ids", "label")
+                    if key in cols_to_retain or key in ("label", "label_ids", "labels")
                 }
+            elif cols_to_retain is not None:
+                batch = {key: value for key, value in batch.items() if key in cols_to_retain}
+
             actual_size = len(list(batch.values())[0])  # Get the length of one of the arrays, assume all same
             # Our collators expect a list of dicts, not a dict of lists/arrays, so we invert
             batch = [{key: value[i] for key, value in batch.items()} for i in range(actual_size)]
@@ -467,23 +499,21 @@ class TensorflowDatasetMixin:
 
         tf_dataset = tf_dataset.map(ensure_shapes)
 
-        if label_cols:
-
-            def split_features_and_labels(input_batch):
-                features = {key: tensor for key, tensor in input_batch.items() if key in columns}
-                labels = {key: tensor for key, tensor in input_batch.items() if key in label_cols}
-                assert set(features.keys()).union(labels.keys()) == set(input_batch.keys())
-                if len(features) == 1:
-                    features = list(features.values())[0]
-                if len(labels) == 1:
-                    labels = list(labels.values())[0]
+        def split_features_and_labels(input_batch):
+            # TODO(Matt, QL): deprecate returning the dict content when there's only one key
+            features = {key: tensor for key, tensor in input_batch.items() if key in columns}
+            labels = {key: tensor for key, tensor in input_batch.items() if key in label_cols}
+            if len(features) == 1:
+                features = list(features.values())[0]
+            if len(labels) == 1:
+                labels = list(labels.values())[0]
+            if isinstance(labels, dict) and len(labels) == 0:
+                return features
+            else:
                 return features, labels
 
+        if cols_to_retain is not None:
             tf_dataset = tf_dataset.map(split_features_and_labels)
-
-        # TODO(Matt, QL): deprecate returning the dict content when there's only one key
-        elif isinstance(tf_dataset.element_spec, dict) and len(tf_dataset.element_spec) == 1:
-            tf_dataset = tf_dataset.map(lambda x: list(x.values())[0])
 
         if prefetch:
             tf_dataset = tf_dataset.prefetch(tf.data.experimental.AUTOTUNE)

--- a/src/datasets/utils/tf_utils.py
+++ b/src/datasets/utils/tf_utils.py
@@ -38,6 +38,14 @@ def minimal_tf_collate_fn(features):
     return batch
 
 
+def minimal_tf_collate_fn_with_renaming(features):
+    batch = minimal_tf_collate_fn(features)
+    if "label" in batch:
+        batch["labels"] = batch["label"]
+        del batch["label"]
+    return batch
+
+
 def is_numeric_pa_type(pa_type):
     if pa.types.is_list(pa_type):
         return is_numeric_pa_type(pa_type.value_type)

--- a/tests/test_arrow_dataset.py
+++ b/tests/test_arrow_dataset.py
@@ -2395,88 +2395,88 @@ class BaseDatasetTest(TestCase):
 
         from datasets.utils.tf_utils import minimal_tf_collate_fn_with_renaming
 
-        with tempfile.TemporaryDirectory() as tmp_dir:
-            with self._create_dummy_dataset(in_memory, tmp_dir, multiple_columns=True) as dset:
-                with dset.rename_columns({"col_1": "features", "col_2": "label"}) as new_dset:
-                    tf_dataset = new_dset.to_tf_dataset(collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4)
-                    batch = next(iter(tf_dataset))
-                    self.assertTrue("labels" in batch and "features" in batch)
+        tmp_dir = tempfile.TemporaryDirectory()
+        with self._create_dummy_dataset(in_memory, tmp_dir.name, multiple_columns=True) as dset:
+            with dset.rename_columns({"col_1": "features", "col_2": "label"}) as new_dset:
+                tf_dataset = new_dset.to_tf_dataset(collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4)
+                batch = next(iter(tf_dataset))
+                self.assertTrue("labels" in batch and "features" in batch)
 
+                tf_dataset = new_dset.to_tf_dataset(
+                    columns=["features", "labels"], collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4
+                )
+                batch = next(iter(tf_dataset))
+                self.assertTrue("labels" in batch and "features" in batch)
+
+                tf_dataset = new_dset.to_tf_dataset(
+                    columns=["features", "label"], collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4
+                )
+                batch = next(iter(tf_dataset))
+                self.assertTrue("labels" in batch and "features" in batch)  # Assert renaming was handled correctly
+
+                with pytest.raises(ValueError):
+                    # Assert that this throws an error when we don't rename columns
                     tf_dataset = new_dset.to_tf_dataset(
-                        columns=["features", "labels"], collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4
-                    )
-                    batch = next(iter(tf_dataset))
-                    self.assertTrue("labels" in batch and "features" in batch)
-
-                    tf_dataset = new_dset.to_tf_dataset(
-                        columns=["features", "label"], collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4
-                    )
-                    batch = next(iter(tf_dataset))
-                    self.assertTrue("labels" in batch and "features" in batch)  # Assert renaming was handled correctly
-
-                    with pytest.raises(ValueError):
-                        # Assert that this throws an error when we don't rename columns
-                        tf_dataset = new_dset.to_tf_dataset(
-                            columns=["features", "label"],
-                            collate_fn=minimal_tf_collate_fn_with_renaming,
-                            batch_size=4,
-                            auto_rename_labels=False,
-                        )
-                        batch = next(iter(tf_dataset))
-
-                    tf_dataset = new_dset.to_tf_dataset(
-                        columns=["features"],
-                        label_cols=["labels"],
-                        collate_fn=minimal_tf_collate_fn_with_renaming,
-                        batch_size=4,
-                    )
-                    batch = next(iter(tf_dataset))
-                    self.assertEqual(len(batch), 2)
-                    # Assert that we don't have any empty entries here
-                    self.assertTrue(isinstance(batch[0], tf.Tensor) and isinstance(batch[1], tf.Tensor))
-
-                    tf_dataset = new_dset.to_tf_dataset(
-                        columns=["features"],
-                        label_cols=["label"],
-                        collate_fn=minimal_tf_collate_fn_with_renaming,
-                        batch_size=4,
-                    )
-                    batch = next(iter(tf_dataset))
-                    self.assertEqual(len(batch), 2)
-                    # Assert that we don't have any empty entries here
-                    self.assertTrue(isinstance(batch[0], tf.Tensor) and isinstance(batch[1], tf.Tensor))
-
-                    with pytest.raises(ValueError):
-                        tf_dataset = new_dset.to_tf_dataset(
-                            columns=["features"],
-                            label_cols=["label"],
-                            collate_fn=minimal_tf_collate_fn_with_renaming,
-                            batch_size=4,
-                            auto_rename_labels=False,
-                        )
-                        batch = next(iter(tf_dataset))
-
-                    tf_dataset = new_dset.to_tf_dataset(
-                        columns=["features"],
-                        collate_fn=minimal_tf_collate_fn_with_renaming,
-                        batch_size=4,
-                        auto_rename_labels=True,
-                    )
-                    batch = next(iter(tf_dataset))
-                    # Assert that labels didn't creep in when we don't ask for them
-                    # just because the collate_fn added them
-                    self.assertTrue(isinstance(batch, tf.Tensor))
-
-                    tf_dataset = new_dset.to_tf_dataset(
-                        columns=["features"],
+                        columns=["features", "label"],
                         collate_fn=minimal_tf_collate_fn_with_renaming,
                         batch_size=4,
                         auto_rename_labels=False,
                     )
                     batch = next(iter(tf_dataset))
-                    # Assert that labels didn't creep in when we don't ask for them
-                    # just because the collate_fn added them
-                    self.assertTrue(isinstance(batch, tf.Tensor))
+
+                tf_dataset = new_dset.to_tf_dataset(
+                    columns=["features"],
+                    label_cols=["labels"],
+                    collate_fn=minimal_tf_collate_fn_with_renaming,
+                    batch_size=4,
+                )
+                batch = next(iter(tf_dataset))
+                self.assertEqual(len(batch), 2)
+                # Assert that we don't have any empty entries here
+                self.assertTrue(isinstance(batch[0], tf.Tensor) and isinstance(batch[1], tf.Tensor))
+
+                tf_dataset = new_dset.to_tf_dataset(
+                    columns=["features"],
+                    label_cols=["label"],
+                    collate_fn=minimal_tf_collate_fn_with_renaming,
+                    batch_size=4,
+                )
+                batch = next(iter(tf_dataset))
+                self.assertEqual(len(batch), 2)
+                # Assert that we don't have any empty entries here
+                self.assertTrue(isinstance(batch[0], tf.Tensor) and isinstance(batch[1], tf.Tensor))
+
+                with pytest.raises(ValueError):
+                    tf_dataset = new_dset.to_tf_dataset(
+                        columns=["features"],
+                        label_cols=["label"],
+                        collate_fn=minimal_tf_collate_fn_with_renaming,
+                        batch_size=4,
+                        auto_rename_labels=False,
+                    )
+                    batch = next(iter(tf_dataset))
+
+                tf_dataset = new_dset.to_tf_dataset(
+                    columns=["features"],
+                    collate_fn=minimal_tf_collate_fn_with_renaming,
+                    batch_size=4,
+                    auto_rename_labels=True,
+                )
+                batch = next(iter(tf_dataset))
+                # Assert that labels didn't creep in when we don't ask for them
+                # just because the collate_fn added them
+                self.assertTrue(isinstance(batch, tf.Tensor))
+
+                tf_dataset = new_dset.to_tf_dataset(
+                    columns=["features"],
+                    collate_fn=minimal_tf_collate_fn_with_renaming,
+                    batch_size=4,
+                    auto_rename_labels=False,
+                )
+                batch = next(iter(tf_dataset))
+                # Assert that labels didn't creep in when we don't ask for them
+                # just because the collate_fn added them
+                self.assertTrue(isinstance(batch, tf.Tensor))
 
         del tf_dataset  # For correct cleanup
 

--- a/tests/test_arrow_dataset.py
+++ b/tests/test_arrow_dataset.py
@@ -2389,6 +2389,98 @@ class BaseDatasetTest(TestCase):
         del tf_dataset  # For correct cleanup
 
     @require_tf
+    def test_tf_label_renaming(self, in_memory):
+        # Protect TF-specific imports in here
+        import tensorflow as tf
+
+        from datasets.utils.tf_utils import minimal_tf_collate_fn_with_renaming
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            with self._create_dummy_dataset(in_memory, tmp_dir, multiple_columns=True) as dset:
+                with dset.rename_columns({"col_1": "features", "col_2": "label"}) as new_dset:
+                    tf_dataset = new_dset.to_tf_dataset(collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4)
+                    batch = next(iter(tf_dataset))
+                    self.assertTrue("labels" in batch and "features" in batch)
+
+                    tf_dataset = new_dset.to_tf_dataset(
+                        columns=["features", "labels"], collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4
+                    )
+                    batch = next(iter(tf_dataset))
+                    self.assertTrue("labels" in batch and "features" in batch)
+
+                    tf_dataset = new_dset.to_tf_dataset(
+                        columns=["features", "label"], collate_fn=minimal_tf_collate_fn_with_renaming, batch_size=4
+                    )
+                    batch = next(iter(tf_dataset))
+                    self.assertTrue("labels" in batch and "features" in batch)  # Assert renaming was handled correctly
+
+                    with pytest.raises(ValueError):
+                        # Assert that this throws an error when we don't rename columns
+                        tf_dataset = new_dset.to_tf_dataset(
+                            columns=["features", "label"],
+                            collate_fn=minimal_tf_collate_fn_with_renaming,
+                            batch_size=4,
+                            auto_rename_labels=False,
+                        )
+                        batch = next(iter(tf_dataset))
+
+                    tf_dataset = new_dset.to_tf_dataset(
+                        columns=["features"],
+                        label_cols=["labels"],
+                        collate_fn=minimal_tf_collate_fn_with_renaming,
+                        batch_size=4,
+                    )
+                    batch = next(iter(tf_dataset))
+                    self.assertEqual(len(batch), 2)
+                    # Assert that we don't have any empty entries here
+                    self.assertTrue(isinstance(batch[0], tf.Tensor) and isinstance(batch[1], tf.Tensor))
+
+                    tf_dataset = new_dset.to_tf_dataset(
+                        columns=["features"],
+                        label_cols=["label"],
+                        collate_fn=minimal_tf_collate_fn_with_renaming,
+                        batch_size=4,
+                    )
+                    batch = next(iter(tf_dataset))
+                    self.assertEqual(len(batch), 2)
+                    # Assert that we don't have any empty entries here
+                    self.assertTrue(isinstance(batch[0], tf.Tensor) and isinstance(batch[1], tf.Tensor))
+
+                    with pytest.raises(ValueError):
+                        tf_dataset = new_dset.to_tf_dataset(
+                            columns=["features"],
+                            label_cols=["label"],
+                            collate_fn=minimal_tf_collate_fn_with_renaming,
+                            batch_size=4,
+                            auto_rename_labels=False,
+                        )
+                        batch = next(iter(tf_dataset))
+
+                    tf_dataset = new_dset.to_tf_dataset(
+                        columns=["features"],
+                        collate_fn=minimal_tf_collate_fn_with_renaming,
+                        batch_size=4,
+                        auto_rename_labels=True,
+                    )
+                    batch = next(iter(tf_dataset))
+                    # Assert that labels didn't creep in when we don't ask for them
+                    # just because the collate_fn added them
+                    self.assertTrue(isinstance(batch, tf.Tensor))
+
+                    tf_dataset = new_dset.to_tf_dataset(
+                        columns=["features"],
+                        collate_fn=minimal_tf_collate_fn_with_renaming,
+                        batch_size=4,
+                        auto_rename_labels=False,
+                    )
+                    batch = next(iter(tf_dataset))
+                    # Assert that labels didn't creep in when we don't ask for them
+                    # just because the collate_fn added them
+                    self.assertTrue(isinstance(batch, tf.Tensor))
+
+        del tf_dataset  # For correct cleanup
+
+    @require_tf
     def test_tf_dataset_options(self, in_memory):
         tmp_dir = tempfile.TemporaryDirectory()
         # Test that batch_size option works as expected

--- a/tests/test_arrow_dataset.py
+++ b/tests/test_arrow_dataset.py
@@ -2436,16 +2436,6 @@ class BaseDatasetTest(TestCase):
                 # Assert that we don't have any empty entries here
                 self.assertTrue(isinstance(batch[0], tf.Tensor) and isinstance(batch[1], tf.Tensor))
 
-                with pytest.raises(ValueError):
-                    tf_dataset = new_dset.to_tf_dataset(
-                        columns=["features"],
-                        label_cols=["label"],
-                        collate_fn=minimal_tf_collate_fn_with_renaming,
-                        batch_size=4,
-                        auto_rename_labels=False,
-                    )
-                    batch = next(iter(tf_dataset))
-
                 tf_dataset = new_dset.to_tf_dataset(
                     columns=["features"],
                     collate_fn=minimal_tf_collate_fn_with_renaming,

--- a/tests/test_arrow_dataset.py
+++ b/tests/test_arrow_dataset.py
@@ -2446,17 +2446,6 @@ class BaseDatasetTest(TestCase):
                 # just because the collate_fn added them
                 self.assertTrue(isinstance(batch, tf.Tensor))
 
-                tf_dataset = new_dset.to_tf_dataset(
-                    columns=["features"],
-                    collate_fn=minimal_tf_collate_fn_with_renaming,
-                    batch_size=4,
-                    auto_rename_labels=False,
-                )
-                batch = next(iter(tf_dataset))
-                # Assert that labels didn't creep in when we don't ask for them
-                # just because the collate_fn added them
-                self.assertTrue(isinstance(batch, tf.Tensor))
-
         del tf_dataset  # For correct cleanup
 
     @require_tf

--- a/tests/test_arrow_dataset.py
+++ b/tests/test_arrow_dataset.py
@@ -2440,7 +2440,6 @@ class BaseDatasetTest(TestCase):
                     columns=["features"],
                     collate_fn=minimal_tf_collate_fn_with_renaming,
                     batch_size=4,
-                    auto_rename_labels=True,
                 )
                 batch = next(iter(tf_dataset))
                 # Assert that labels didn't creep in when we don't ask for them

--- a/tests/test_arrow_dataset.py
+++ b/tests/test_arrow_dataset.py
@@ -2414,16 +2414,6 @@ class BaseDatasetTest(TestCase):
                 batch = next(iter(tf_dataset))
                 self.assertTrue("labels" in batch and "features" in batch)  # Assert renaming was handled correctly
 
-                with pytest.raises(ValueError):
-                    # Assert that this throws an error when we don't rename columns
-                    tf_dataset = new_dset.to_tf_dataset(
-                        columns=["features", "label"],
-                        collate_fn=minimal_tf_collate_fn_with_renaming,
-                        batch_size=4,
-                        auto_rename_labels=False,
-                    )
-                    batch = next(iter(tf_dataset))
-
                 tf_dataset = new_dset.to_tf_dataset(
                     columns=["features"],
                     label_cols=["labels"],

